### PR TITLE
Enable multiple registrations

### PR DESCRIPTION
**Pull request by dpanta94**
The primary funcitonality that is being introduced is allowing you to make multiple registrations just by using the API exposed.

Previously, because internally there were calls to wp_script_is or wp_style_is even if you removed the asset altogether you could not register more than once.

Registering more than once can have some advantages - listing a couple that pop on the top of my head.

1. Adding a script which is extending another script related to block editor.

Now you can modify your original's script dependencies to include your new script. Making managing when your script should be included easy. Wherever my original script is included, also please include my extension.

2. Enabling us to overwrite based on filters.

Also this PR introduces a magic method which enables us to call script or style methods from WP on the asset instance itself. You only need to call the method without the `wp_` prefix and replacing with `asset` wherever the `script` or `style` would be.

So now you can do:

```php
$asset->deregister_asset();

// will correctly call wp_deregister_script( $slug ) or wp_deregister_style( $slug ), based on the asset's type.
// The slug is always passed as the first param and you can pass anything else you might need.
```

See an example of a snippet we had to provide support to achieve the functionality being added here:
https://lw.slack.com/archives/C01SBD5T03V/p1732799298015619?thread_ts=1732270965.841129&cid=C01SBD5T03V

### CODE REVIEW

File: src/Assets/Asset.php [hunk 1]
```diff
@@ -1692,27 +1700,142 @@ public function set_as_registered() {
...
+	 * on the asset.
+	 *
+	 * example: $asset->register_asset( ...$args ); will call wp_register_script( $slug, ...$args ); or
+	 * wp_register_style( $slug, ...$args ); based on if the asset is JS or CSS type.
+	 *
+	 * @since TBD
+	 *
+	 * @return static
+	 */
+	public function __call( $method, $args ) {
```
Comment (lucatume) [inline src/Assets/Asset.php:1714]: This is smart and saves some lines of code, but I do not think makes for readable code.
I would unroll the `__call` method and explicitly create each method that is now part of the `@method` annotation.
This should likely not be a gateway to **all** `wp_` functions, but to a limited, controlled and **tested** set of them.
What would this do if I call `$asset->footer`?
If you end up having to maintain a safe-list of functions that can be called, you're better served by an explicit API made of concrete methods.

**Review by lucatume — CHANGES_REQUESTED** [review]

**Comment by borkweb** [comment]
This looks pretty great (once Luca's comment is addressed) 🕺

File: src/Assets/Asset.php [hunk 2]
```diff
@@ -1696,27 +1699,190 @@ public function set_as_registered() {
...
+	public function register_asset( string $url = '', array $dependencies = [], string $version = null, $in_footer_or_media = 'all' ) {
+		$url = $url ? $url : $this->get_url();
+		$dependencies = $dependencies ? $dependencies : $this->get_dependencies();
+		$version = $version ?? $this->get_version();
+		$in_footer_or_media = $in_footer_or_media ? $in_footer_or_media : ( $this->is_js() ? $this->is_in_footer() : $this->get_media() );
+
+		$method = 'wp_register_' . $this->get_script_or_style();
+		$method( $this->get_slug(), $url, $dependencies, $version, $in_footer_or_media );
+		return $this;
+	}
```
Comment (borkweb) [inline src/Assets/Asset.php:1749]: This feels redundant with `register()`. If we really need this logic in here, is it best to replace the functionality of `register()`? To skip around that method sorta breaks the library's purpose.

**Review by borkweb — CHANGES_REQUESTED** [review]
I apologize that I did not think/catch this sooner, but I think this needs a bit more discussion/thought. I'm happy to jump on a call next week if it'd be helpful to talk it out.

Right now, as I re-review it a second time, my spidey senses are tingling - probably based on the unfurling of that `__call()` method and it trigging the appropriate level of attention that I was missing before.

## Question A: Should an asset know how to output itself, dequeue itself, etc?

Probably. I don't dislike the movement of some of that awareness and execution into the `Asset` class itself. However, it does feel a bit off to have the bulk of the orchestration remain in the `Assets` library while just pieces going to the `Asset` library.

## Question B: Is there a deeper reason for the redundant methods being created that I'm not understanding?

The simple answer to this is probably **yes**, but as of right now I'm not following. I think seeing `register()`, `do_register()`, and `register_asset()` is an incredibly confusing thing to someone without them digging into the code to understand why there's a distinction there. I, personally, don't know why we've added the two extra methods.

If we do want to have the asset know how to register itself in WP; know how to enqueue itself in WP; etc, those should be done in `protected` methods so the API methods that users interact with are restricted to `enqueue()`, `dequeue()`, and `register()`.

## Question C: Can you share the thing this is trying to solve in a video walkthrough or something or in a call with me?

I believe I am misunderstanding the purpose. Is the goal to be able to allow the following?

```php
Asset::add( 'thing', /* whatever */ )
  ->register();

// Do some stuff.

Assets::get( 'thing' )
  ->dequeue()
  ->add_dependency( 'kdfdfdfdf' );
  ->do_enqueue();
```

That's the only thing I can think of. _If_ that is the case (which I'm guessing it is not), I'm not a huge fan of that and I think there's other ways to accomplish that sort of thing.

## TL;DR:

I believe I'm misunderstanding the goal that this changeset aims to accomplish now that I've reviewed it a second time. I'd love to chat!
